packages/vscode-extension/src/utilities/merge.ts

```diff
@@ -0,0 +1,46 @@
+import { RecursivePartial } from "../common/State";
+
+/**
+ * Merges an existing state object with updates from a new partial state.
+ *
+ * @param {T} oldNode - The original state object.
+ * @param {RecursivePartial<T>} newNode - A partial state object containing updates.
+ *
+ * @returns {T} The new state object post-merge.
+ */
+export function merge<T extends { [P in keyof T]: T[P] }>(
+  oldNode: T,
+  newNode: RecursivePartial<T>
+): T {
+  const result: RecursivePartial<T> = {};
+  let wasChanged = false;
+
+  const allKeys = [...Object.keys(oldNode), ...Object.keys(newNode)] as (keyof T)[];
+
+  for (const key of allKeys) {
+    if (newNode[key] === undefined || newNode[key] === null) {
```

Comment:
I think we shouldn't special-case `null`, it can be useful to represent "empty" states.